### PR TITLE
chainntnfs: fix nil pointer on address reuse

### PR DESCRIPTION
**Pull request by guggero**
When a confirmation notification for a script hash is registered, the notifier panics with a nil pointer dereference in case two transactions send funds to this script and are confirmed at different block heights.

The test that triggers the panic is added in the first commit with a fix for it in the second commit.

### CODE REVIEW

File: chainntnfs/txnotifier_test.go [hunk 1]
```diff
@@ -604,6 +605,112 @@ func TestTxNotifierFutureSpendDispatch(t *testing.T) {
 	}
 }
 
+// TestTxNotifierFutureConfDispatchReuseSafe tests that the notifier does not
```
[the remark this answers is not in the document]
Comment (guggero) [reply chainntnfs/txnotifier_test.go]: Right, makes sense.

File: chainntnfs/txnotifier_test.go [hunk 2]
```diff
@@ -604,6 +605,112 @@ func TestTxNotifierFutureSpendDispatch(t *testing.T) {
...
+			t.Fatalf("unable to connect block: %v", err)
+		}
+		if err := n.NotifyHeight(currentBlock); err != nil {
+			t.Fatalf("unable to dispatch notifications: %v", err)
+		}
+	}
+
+	// If we reach this point without panic, the bug this test was written
+	// for no longer exists.
+	if !status1.confirmed || !status2.confirmed {
```
[the remark this answers is not in the document]
Comment (guggero) [reply chainntnfs/txnotifier_test.go]: Yes, I just saw the race test failure today.
We need to read multiple events from the channels because there will be multiple confirmations the way it is implemented at the moment. So I solved it with wait groups for now.

But maybe that's another reason to go with the other implementation where we clean out all scripts that have been confirmed and therefore would only send one confirmation event.

[the remark this answers is not in the document]
Comment (halseth) [reply chainntnfs/txnotifier_test.go]: Yeah, I think that's the way it should be fixed.

**Comment by guggero** [comment]
I now implemented the version that will not notify again on reuse.
Turns out, it's also less code.
Tested it with my regtest setup and was unable to reproduce it after the fix.

The unit test setup is still a bit peculiar with the event listener in the goroutine since we still get multiple messages on the `Event.Update` channel.
The reason for this is that because both `confRequest`s are identical, there are two `ntfns` created for the same request and they both get notified. And since the request is referenced from two block heights, we get a total of 4 `Update` messages (two per TX).
And because the chan buffer is only 1, the test would block if we didn't read in a goroutine.

Not sure if this can still cause problems on the receiver end?

**Comment by wpaulino** [comment]
> The reason for this is that because both confRequests are identical, there are two ntfns created for the same request and they both get notified. And since the request is referenced from two block heights, we get a total of 4 Update messages (two per TX).

I think we can get around this by adding an additional check to ensure no details exist for the notification before notifying the transaction at:
https://github.com/lightningnetwork/lnd/blob/6765b8668a2a08f58b38480fff3534fcfa4c001e/chainntnfs/txnotifier.go#L1494

**Comment by guggero** [comment]
I was able to get around the problem by not notifying `ntfn`s that have already been marked as `dispatched`. Now the test is simpler too.

**Comment by guggero** [comment]
Ok, I think I was finally able to fix it the way you suggested, @wpaulino.

File: chainntnfs/txnotifier.go [hunk 3]
```diff
@@ -633,7 +633,34 @@ func (n *TxNotifier) RegisterConf(txid *chainhash.Hash, pkScript []byte,
 		confSet = newConfNtfnSet()
 		n.confNotifications[ntfn.ConfRequest] = confSet
 	}
-	confSet.ntfns[ntfn.ConfID] = ntfn
+
+	// In case of an address reuse, we would get multiple confirmation
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply chainntnfs/txnotifier.go]: Yeah that's what I also had in mind, adding the small check in `handleConfDetailsAtTip`. I don't think any changes need to be done for registrations, the assumption there is that it'll dispatch on the first transaction that satisfies it starting from the initial height hint.

[the remark this answers is not in the document]
Comment (guggero) [reply chainntnfs/txnotifier.go]: Ok, it would have been very simple if I wasn't looking in the wrong place the whole time... Sorry about that.
This indeed doesn't make it necessary to change anything with the registrations.

File: chainntnfs/txnotifier_test.go [hunk 4]
```diff
@@ -604,6 +604,112 @@ func TestTxNotifierFutureSpendDispatch(t *testing.T) {
...
+	}
+	select {
+	case <-ntfn1.Event.Confirmed:
+	default:
+		t.Fatal("expected confirmation of TX 1")
+	}
+
+	// Now register a second TX that spends to the same output and get it
+	// confirmed. This TX needs to have a different TX hash, otherwise
+	// we'd get strange results. So we tweak one field in the TxIn.
```
Comment (wpaulino) [inline chainntnfs/txnotifier_test.go]: Does this comment still apply? `tx2` looks the same as `tx1`. I think we do want different txids here.

File: chainntnfs/txnotifier.go [hunk 5]
```diff
@@ -1507,6 +1507,13 @@ func (n *TxNotifier) handleConfDetailsAtTip(confRequest ConfRequest,
 
 	// TODO(wilmer): cancel pending historical rescans if any?
 	confSet := n.confNotifications[confRequest]
+	
+	// If we already have details for this request, we don't want to add it
+	// again since we have already dispatched notifications for it.
+	if confSet.details != nil {
```
Comment (wpaulino) [inline chainntnfs/txnotifier.go]: Commit message needs to be rewritten for this change.

File: chainntnfs/txnotifier_test.go (hunk 4 above)
Comment (guggero) [reply chainntnfs/txnotifier_test.go]: Comment was wrong, we want the same TX. Fixed.

File: chainntnfs/txnotifier.go (hunk 5 above)
Comment (guggero) [reply chainntnfs/txnotifier.go]: Done.

File: chainntnfs/txnotifier.go [hunk 6]
```diff
@@ -1507,6 +1507,13 @@ func (n *TxNotifier) handleConfDetailsAtTip(confRequest ConfRequest,
 
 	// TODO(wilmer): cancel pending historical rescans if any?
 	confSet := n.confNotifications[confRequest]
+
+	// If we already have details for this request, we don't want to add it
+	// again since we have already dispatched notifications for it.
+	if confSet.details != nil {
+		return
```
Comment (halseth) [inline chainntnfs/txnotifier.go]: add a logging statement?

File: chainntnfs/txnotifier_test.go [hunk 7]
```diff
@@ -604,6 +604,110 @@ func TestTxNotifierFutureSpendDispatch(t *testing.T) {
...
+	case <-ntfn1.Event.Confirmed:
+	default:
+		t.Fatal("expected confirmation of TX 1")
+	}
+
+	// Now register a second TX that spends to the same output.
+	tx2 := wire.MsgTx{Version: 1}
+	tx2.AddTxOut(&wire.TxOut{PkScript: testRawScript})
+	tx2Hash := tx2.TxHash()
+	ntfn2, err := n.RegisterConf(&tx2Hash, testRawScript, 1, 1)
```
Comment (halseth) [inline chainntnfs/txnotifier_test.go]: We should also test registering with only the script (no txid). In that case we expect to receive confirmation only for the first tx I think, shich we can check by inspecting the `TxConfirmtion`

File: chainntnfs/txnotifier.go [hunk 8]
```diff
@@ -1507,6 +1507,13 @@ func (n *TxNotifier) handleConfDetailsAtTip(confRequest ConfRequest,
 
 	// TODO(wilmer): cancel pending historical rescans if any?
 	confSet := n.confNotifications[confRequest]
+
+	// If we already have details for this request, we don't want to add it
+	// again since we have already dispatched notifications for it.
+	if confSet.details != nil {
```
Comment (halseth) [inline chainntnfs/txnotifier.go]: Maybe also update the godoc for the `detail` field as it is becoming more than just a cache at this point

File: chainntnfs/txnotifier.go (hunk 6 above)
Comment (guggero) [reply chainntnfs/txnotifier.go]: Done. Hope the wording and log level is adequate.

File: chainntnfs/txnotifier_test.go (hunk 7 above)
Comment (guggero) [reply chainntnfs/txnotifier_test.go]: Yes, definitely. Added the additional checks.

File: chainntnfs/txnotifier_test.go [hunk 9]
```diff
@@ -604,6 +604,157 @@ func TestTxNotifierFutureSpendDispatch(t *testing.T) {
...
+	case details := <-scriptNtfn1.Event.Confirmed:
+		assertConfDetails(t, details, confDetails)
+	default:
+		t.Fatal("expected update of script ntfn")
+	}
+
+	// Now register a second TX that spends to the same output and again
+	// register a confirmation for just the script.
+	tx2 := wire.MsgTx{Version: 1}
+	tx2.AddTxOut(&wire.TxOut{PkScript: testRawScript})
```
Comment (wpaulino) [inline chainntnfs/txnotifier_test.go]: Nit: maybe add another output to the transaction so that it results in a different txid?

File: chainntnfs/txnotifier.go [hunk 10]
```diff
@@ -1507,6 +1510,17 @@ func (n *TxNotifier) handleConfDetailsAtTip(confRequest ConfRequest,
...
+		Log.Infof("Not registering another notification for %v. "+
```
Comment (wpaulino) [inline chainntnfs/txnotifier.go]: Nit: would just reword this to `Ignoring address reuse for <script> at height <height>`.
